### DISCUSSION TITLE
Updating ui components

### DISCUSSION THREAD

**Discussion by AndiLeni** [zauberzeug/nicegui#590] [Q&A] [answered]
Hello,

I have some issues with updating ui components.
I need to update a label after each iteration for my ai algorithm.
Unfortunately, the label is only updated visually after the complete learning cycle is finished (event thou update_label() is called in each iteration)

Is there any way to somehow achieve this and force a ui update?

Thanks and kind regards

This is a minimal example of my situation:

main.py
```py
from Page import Page
from nicegui import ui


@ui.page("/")
def p():
    Page()    

ui.run()
```

Page.py
```py
from nicegui import ui
from Sec import Sec

class Page:
    def __init__(self) -> None:
       self.label = None
       self.sec = Sec()

       self.create_page()
    
    def update_label(self, value):
        print(f"updating label with value: {value}")
        self.label.set_text(f"iteration {value}")

    def start(self):
        print("starting iteration")
        self.sec.iter(gui=self)
        
    def create_page(self):
        self.label = ui.label("Hello NiceGUI!")
        ui.button('Start', on_click=self.start)
```

Sec.py
```py
import time

class Sec:
    def __init__(self) -> None:
        pass

    def iter(self, gui):
        for i in range(1, 5):
            print(f"iteration {i}")
            gui.update_label(i)
            time.sleep(1)
```

**Comment by falkoschindler** [comment, marked as answer]
The problem is that your computation runs synchronously. There is no way for NiceGUI to do anything UI-related in between, so the client is not updated.

This is demonstrated with the following simple example. You only see "Done!" once all computation steps are completed:

```py
def compute():
    for i in range(100):
        computation_step()
        status.text = f'{i} %'
    status.text = 'Done!'

def computation_step():
    time.sleep(0.01)

status = ui.label()
ui.button('Start', on_click=compute)
```

Instead, you should run computation-heavy functions in an "executor" to be able to await them. Every time your code reaches an `await`, other tasks that have been waiting can take their turn. Updating the UI is such a task. Now you see the percentage increasing:

```py
async def compute():
    for i in range(100):
        await asyncio.get_event_loop().run_in_executor(None, computation_step)
        status.text = f'{i} %'
    status.text = 'Done!'

def computation_step():
    time.sleep(0.01)

status = ui.label()
ui.button('Start', on_click=compute)
```

**Reply by AndiLeni** [reply]
Thanks for your suggestion, I will give this a try :-)
